includes/Core/Modules/Module.php

```diff
@@ -461,6 +487,65 @@ public function is_connected() {
 		return $this->parse_data_response( $data, $response );
 	}
 
+	/**
+	 * Validates the given data request.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Data_Request $data Data request object.
+	 * @throws Insufficient_Scopes_Exception Thrown if the user has not granted
+	 *                                       necessary scopes required by the datapoint.
+	 */
+	private function validate_data_request( Data_Request $data ) {
+		$definitions   = $this->get_datapoint_definitions();
+		$datapoint_key = "$data->method:$data->datapoint";
+
+		// All datapoints must be defined.
+		if ( empty( $definitions[ $datapoint_key ] ) ) {
+			throw new Exception( __( 'Invalid datapoint.', 'google-site-kit' ), 400 );
+		}
+
+		if ( empty( $definitions[ $datapoint_key ]['scopes'] ) ) {
+			return;
+		}
+
+		$datapoint = $definitions[ $datapoint_key ];
+
+		// If the datapoint requires specific scopes, ensure they are satisfied.
+		if ( ! $this->authentication->get_oauth_client()->has_sufficient_scopes( $datapoint['scopes'] ) ) {
+			$exception = new Insufficient_Scopes_Exception( $datapoint['request_scopes_message'] );
```

Comment:
We _may_ be better off separating the `error.message` from `error.data.requestScopesMessage` in JS, as originally in the IB (right now the error as implemented here will only include `error.message`). Depending on what messages we're going to use, what's logged (and potentially displayed) as general error messages may not be a good message to present to the user, and vice-versa. We are effectively "catching" the error in JS and then showing a modal with a more actionable explanation, so we could argue this should not be the same error message that should have been displayed if we didn't build the client-side mechanism.

Reply:
I understand there is a potential benefit of separating an error message from a user-facing display message. I don't think we do that anywhere else and caught exceptions are turned into user-facing `WP_Error`s with the result of `getMessage()` as the message users will see. I'm open to ideas here, but it seems to me that the `$message` is intended to be user-facing as the class, `code`, and other data is more relevant for internal use.

Reply:
That's fair, let's go with this for now.